src/couch_index_server.erl

```diff
@@ -131,50 +144,111 @@ terminate(_Reason, _State) ->
+maybe_close_idle(#st{open=Open, soft_max_indexes=Max}=State) when Open < Max ->
+    {ok, State};
+
+maybe_close_idle(State) ->
+    ets:safe_fixtable(?BY_IDLE, true),
+    try
+        {ok, close_idle(State, ets:first(?BY_IDLE))}
+    after
+        ets:safe_fixtable(?BY_IDLE, false)
+    end.
+
+
+-spec close_idle(#st{}, term()) -> #st{}.
+close_idle(State, '$end_of_table') ->
+    State;
+
+close_idle(State, {DbName, Sig}) ->
+    true = ets:delete(?BY_IDLE, {DbName, Sig}),
+    case ets:lookup(?BY_SIG, {DbName, Sig}) of
+        [{_, {Pid, _Monitor}}] ->
+            couch_index:stop(Pid),
+            [{DbName, {DDocId, Sig}}] =
+                ets:match_object(?BY_DB, {DbName, {'$1', Sig}}),
+            rem_from_ets(DbName, Sig, DDocId, Pid),
+            case closed(State) of
+                NewState when NewState#st.open > State#st.soft_max_indexes ->
+                    close_idle(NewState, ets:next(?BY_IDLE, {DbName, Sig}));
+                NewState ->
+                    NewState
+            end;
+        [] ->
+            close_idle(State, ets:next(?BY_IDLE, {DbName, Sig}))
+    end.
+
+
 handle_call({get_index, {_Mod, _IdxState, DbName, Sig}=Args}, From, State) ->
     case ets:lookup(?BY_SIG, {DbName, Sig}) of
         [] ->
+            {ok, NewState} = maybe_close_idle(State),
             spawn_link(fun() -> new_index(Args) end),
-            ets:insert(?BY_SIG, {{DbName, Sig}, [From]}),
-            {noreply, State};
-        [{_, Waiters}] when is_list(Waiters) ->
-            ets:insert(?BY_SIG, {{DbName, Sig}, [From | Waiters]}),
+            Monitor = couch_index_monitor:spawn_link({DbName, Sig}),
+            ets:insert(?BY_SIG, {{DbName, Sig}, {[From], Monitor}}),
+            {noreply, NewState};
+        [{_, {Waiters, Monitor}}] when is_list(Waiters) ->
+            ets:insert(?BY_SIG, {{DbName, Sig}, {[From | Waiters], Monitor}}),
             {noreply, State};
-        [{_, Pid}] when is_pid(Pid) ->
+        [{_, {Pid, Monitor}}] when is_pid(Pid) ->
+            ok = incref({DbName, Sig}),
+            ok = couch_index_monitor:notify(Monitor, From),
             {reply, {ok, Pid}, State}
     end;
 handle_call({async_open, {DbName, DDocId, Sig}, {ok, Pid}}, _From, State) ->
-    [{_, Waiters}] = ets:lookup(?BY_SIG, {DbName, Sig}),
-    [gen_server:reply(From, {ok, Pid}) || From <- Waiters],
+    [{_, {Waiters, Monitor}}] = ets:lookup(?BY_SIG, {DbName, Sig}),
     link(Pid),
-    add_to_ets(DbName, Sig, DDocId, Pid),
-    {reply, ok, State};
-handle_call({async_error, {DbName, _DDocId, Sig}, Error}, _From, State) ->
-    [{_, Waiters}] = ets:lookup(?BY_SIG, {DbName, Sig}),
+    ets:insert(?BY_SIG, {{DbName, Sig}, {Pid, Monitor}}),
+    ets:insert(?BY_PID, {Pid, {DbName, Sig}}),
+    ets:insert(?BY_COUNTERS, {{DbName, Sig}, 0}),
+    ets:insert(?BY_DB, {DbName, {DDocId, Sig}}),
+    lists:foreach(fun(From) ->
+        {Client, _} = From,
+        ok = incref({DbName, Sig}),
```

Comment:
The reason behind manual ref counting is because it eliminates races while still not centralizing the "get already-open index" and "mark as idle" logic. So, the only logic which is centralized in couch_index_server is the "open currently-closed index" and "close idle index" logic.